sklearn/utils/estimator_checks.py

```diff
@@ -468,8 +469,61 @@ def check_dict_unchanged(name, Estimator):
                               'Estimator changes __dict__ during %s' % method)
 
 
+def is_public_parameter(attr):
+    return not (attr.startswith('_') or attr.endswith('_'))
+
+
+def check_dont_overwrite_parameters(name, Estimator):
+    # check that fit method only changes or sets private attributes
+    if hasattr(Estimator.__init__, "deprecated_original"):
+        # to not check deprecated classes
+        return
+    rnd = np.random.RandomState(0)
+    X = 3 * rnd.uniform(size=(20, 3))
+    y = X[:, 0].astype(np.int)
+    y = multioutput_estimator_convert_y_2d(name, y)
+    estimator = Estimator()
+    set_testing_parameters(estimator)
+
+    if hasattr(estimator, "n_components"):
+        estimator.n_components = 1
+    if hasattr(estimator, "n_clusters"):
+        estimator.n_clusters = 1
+
+    set_random_state(estimator, 1)
+    dict_before_fit = estimator.__dict__.copy()
+    estimator.fit(X, y)
+
+    dict_after_fit = estimator.__dict__
+
+    public_keys_after_fit = [key for key in dict_after_fit.keys()
+                             if is_public_parameter(key)]
+
+    attrs_added_by_fit = [key for key in public_keys_after_fit
+                          if key not in dict_before_fit.keys()]
+
+    # check that fit doesn't add any public attribute
+    assert_true(not attrs_added_by_fit,
+                ('Estiamtor adds public attribute(s) during'
+                 ' the fit method.'
+                 ' Estimators are only allowed to add private attributes'
+                 ' either started with _ or ended'
+                 ' with _ but %s added' % ', '.join(attrs_added_by_fit)))
+
+    # check that fit doesn't change any public attribute
+    attrs_changed_by_fit = [key for key in public_keys_after_fit
+                            if (dict_before_fit[key] != dict_after_fit[key])]
```

Comment:
Actually, in some ways it's a weaker test, in other ways an unnecessarily strengthened test :\